### PR TITLE
openstack: Generate cloud provider config

### PR DESCRIPTION
**Pull request by Fedosin**
For in-tree cloud provider we need ini-based config that is stored in a secret. To do so we add we write information about the secret in the local config file (/etc/kubernetes/cloud.conf) and store the real config in the secret.

### CODE REVIEW

**Comment by trown** [comment]
/test e2e-openstack

openstack CI should be fixed

**Comment by tomassedovic** [comment]
This is an important piece to enable the OpenStack cloud provider. I've tested this on OpenStack and I'm happy with the PR, but it touches stuff outside of that, too.

These two PRs will be needed to enable the cloud provider end to end, but they don't block this merge:

https://github.com/openshift/origin/pull/22799
https://github.com/openshift/machine-config-operator/pull/661/

@wking any isuggestion for a good person to review this? Presumably we should hold this until 4.2 master opens?

**Comment by patrickdillon** [comment]
I think scaleup test is broken. Feel free to skip it while we debug.

**Comment by derekwaynecarr** [comment]
the change makes sense to me.

nodes need read access to the secret, and the file on disk doesn't contain any secret information.

@abhinavdahiya can this get review?

File: pkg/asset/manifests/template.go [hunk 1]
```diff
@@ -19,7 +19,8 @@ type AzureCredsSecretData struct {
 
 // OpenStackCredsSecretData holds encoded credentials and is used to generate cloud-creds secret
 type OpenStackCredsSecretData struct {
-	Base64encodeCloudCreds string
+	Base64encodeCloudCreds    string
+	Base64encodeCloudCredsINI string
```
Comment (patrickdillon) [inline pkg/asset/manifests/template.go]: nit: I think encode should be capitalized. `Base64encodeCloudCreds` -> `Base64EncodeCloudCreds` for both field names

File: pkg/asset/templates/content/openshift/rolebinding-cloud-creds-secret-reader.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+const (
+	rolebindingCloudCredsSecretReaderFileName = "rolebinding-cloud-creds-secret-reader.yaml.template"
+)
+
+var _ asset.WritableAsset = (*RolebindingCloudCredsSecretReader)(nil)
+
+// RolebindingCloudCredsSecretReader is the variable to represent contents of corresponding file
+type RolebindingCloudCredsSecretReader struct {
```
Comment (patrickdillon) [inline pkg/asset/templates/content/openshift/rolebinding-cloud-creds-secret-reader.go]: nit: throughout this file I think binding should be capitalized `Rolebinding` -> `RoleBinding`

File: pkg/asset/manifests/template.go (hunk 1 above)
Comment (Fedosin) [reply pkg/asset/manifests/template.go]: Sure, will be done!

File: pkg/asset/templates/content/openshift/rolebinding-cloud-creds-secret-reader.go (hunk 2 above)
Comment (Fedosin) [reply pkg/asset/templates/content/openshift/rolebinding-cloud-creds-secret-reader.go]: ditto

File: pkg/asset/manifests/template.go (hunk 1 above)
Comment (Fedosin) [reply pkg/asset/manifests/template.go]: Actually I just found that fields of all cloud providers are called in the same manner. Maybe we should just leave them as they are to keep them consistent?https://github.com/openshift/installer/blob/master/pkg/asset/manifests/template.go#L4-L18

File: data/data/manifests/openshift/cloud-creds-secret.yaml.template [hunk 3]
```diff
@@ -25,6 +25,7 @@ data:
   azure_region: {{.CloudCreds.Azure.Base64encodeRegion}}
 {{- else if .CloudCreds.OpenStack}}
   clouds.yaml: {{.CloudCreds.OpenStack.Base64encodeCloudCreds}}
+  clouds.conf: {{.CloudCreds.OpenStack.Base64encodeCloudCredsINI}}
```
Comment (abhinavdahiya) [inline data/data/manifests/openshift/cloud-creds-secret.yaml.template]: Can you provide details on why we are including the `clouds.conf` in the secret that is used to talk to the API...

`clouds.conf` should be available from `openshift-config/cloud-provider-config` config map... ?

File: pkg/asset/manifests/openstack/cloudproviderconfig.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
+package openstack
+
+import (
+	"bytes"
+
+	"github.com/gophercloud/utils/openstack/clientconfig"
+	"github.com/pkg/errors"
+
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/openstack/cloudproviderconfig.go]: extra line.

File: pkg/asset/manifests/openstack/cloudproviderconfig.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	DomainID   string `ini:"domain-id,omitempty"`
+	DomainName string `ini:"domain-name,omitempty"`
+	Region     string `ini:"region,omitempty"`
+	CAFile     string `ini:"ca-file,omitempty"`
+}
+
+// CloudProviderConfig generates the cloud provider config for the OpenStack platform.
+func CloudProviderConfig() string {
+	return `[Global]
+secret-name = openstack-credentials
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/openstack/cloudproviderconfig.go]: we should work towards moving to a separate secret, as the `kube-system/openstack-credentials` is used by creds-operator to provide creds to operators to talk to API, and this change overloads it to some degree.. Bu it is fine for now.

**Pull request merged**